engine/src/test/scala/cromwell/CromwellTestkitSpec.scala

```diff
@@ -341,25 +344,40 @@ abstract class CromwellTestkitSpec extends TestKit(new CromwellTestkitSpec.TestW
                             (implicit ec: ExecutionContext): WorkflowId = {
     val workflowManager = workflowManagerActor.getOrElse(buildWorkflowManagerActor(config))
     val sources = sampleWdl.asWorkflowSources(runtime, workflowOptions)
+    val maxRetries = 3
+    def isFatal(e: Throwable) = e match {
+      case _: OutputNotFoundException => false
+      case _ => true
+    }
     eventFilter.intercept {
       within(timeoutDuration) {
         val workflowId = workflowManager.submit(sources)
         verifyWorkflowState(workflowManager, workflowId, terminalState)
-        val outputs = getWorkflowOutputsFromMetadata(workflowId)
-        val actualOutputNames = outputs.keys mkString ", "
-        val expectedOutputNames = expectedOutputs.keys mkString " "
 
-        expectedOutputs foreach { case (outputFqn, expectedValue) =>
-          val actualValue = outputs.getOrElse(outputFqn, throw new RuntimeException(s"Expected output $outputFqn was not found in: '$actualOutputNames'"))
-          if (expectedValue != AnyValueIsFine) validateOutput(actualValue, expectedValue)
-        }
-        if (!allowOtherOutputs) {
-          outputs foreach { case (actualFqn, actualValue) =>
-            val expectedValue = expectedOutputs.getOrElse(actualFqn, throw new RuntimeException(s"Actual output $actualFqn was not wanted in '$expectedOutputNames'"))
+        def checkOutputs() = Future {
+          val outputs = getWorkflowOutputsFromMetadata(workflowId)
+          val actualOutputNames = outputs.keys mkString ", "
+          val expectedOutputNames = expectedOutputs.keys mkString " "
+
+          expectedOutputs foreach { case (outputFqn, expectedValue) =>
+            val actualValue = outputs.getOrElse(outputFqn, throw new OutputNotFoundException(outputFqn, actualOutputNames))
             if (expectedValue != AnyValueIsFine) validateOutput(actualValue, expectedValue)
           }
+          if (!allowOtherOutputs) {
+            outputs foreach { case (actualFqn, actualValue) =>
+              val expectedValue = expectedOutputs.getOrElse(actualFqn, throw new RuntimeException(s"Actual output $actualFqn was not wanted in '$expectedOutputNames'"))
+              if (expectedValue != AnyValueIsFine) validateOutput(actualValue, expectedValue)
+            }
+          }
+          workflowId
         }
-        workflowId
+
+        // Retry because we have no guarantee that all the metadata is there when the workflow finishes...
+        Await.result(Retry.withRetry(
+          checkOutputs,
+          maxRetries = Some(maxRetries),
+          isFatal = isFatal,
+          backoff = SimpleExponentialBackoff(0.5 seconds, 0.5 second, 1D)), timeoutDuration)
```

Comment:
This seems like it would be a generally useful bit of code for our brave new eventually consistent world. Specifically, I'm going to need something like this on my query API branch soon. 😄